### ISSUE TITLE
Mod not starting on server

### ISSUE THREAD

**Issue by dann1kid** [ApacheTechSolutions/ApacheTech.VintageMods.CampaignCartographer#14]
**Describe the Bug:**
mod not starting on mp server

**To Reproduce:**
install mod to mods folder on 1.16.1 mp server and start server

**Expected Behaviour:**
mod loads and work

**Additional Context**
logs
```
[Error] [campaigncartographer] An exception was thrown when trying to start the mod:
System.NullReferenceException: Ссылка на объект не указывает на экземпляр объекта.
   в System.Object.GetType()
   в ApacheTech.Common.DependencyInjection.ServiceCollection.RegisterSingleton[TService](TService instance)
   в ApacheTech.VintageMods.Core.Hosting.DependencyInjection.Registration.ServerApiRegistrar.RegisterServerApiEndpoints(IServiceCollection container)
   в ApacheTech.Common.DependencyInjection.Extensions.HostExtensions.Configure(IServiceCollection services, Action`1 serviceRegistrationFactory)
   в ApacheTech.VintageMods.Core.Hosting.ModHost.BuildServerHost(ICoreServerAPI sapi)
   в ApacheTech.VintageMods.Core.Hosting.ModHost.StartPre(ICoreAPI api)
   в Vintagestory.Common.ModLoader.TryStartSystem(Mod mod, ModSystem system, ICoreAPI api, ModEventPhase phase)```

**Comment by ApacheTech** [comment]
Can I ask if this is a recurring issue, or a single event?

I've made changes to ensure this doesn't happen again. It is very surprising that this should be an issue though. It'd be interesting to see the full logs from this crash.

**Comment by ApacheTech** [comment]
**Developer Notes:**

It's hard to see how this crash could have occurred, short of a major blip with the server. It was caused by the lazy-loading of the MEF Mediator, on the server side. This is the only entry in the server-side IOC container. It is possible that a double registration occurred, because the mediator will have been registered as a ModSystem in its own right; however, this shouldn't have been an issue at all, and as the ModSystem is eager-loaded, that would have crashed beforehand, if it was a recurring problem.

For now, I have disabled the MEF Mediator entirely, as it's not currently used. This is worth considering for future releases. Stress tests should include multiple mods, power loss, process suspension, dramatic lag, and unexpected closure of server.

**Comment by dann1kid** [comment]
I not checked it twice after cashe clearing (what very helpful). Write later after retest

**Comment by dann1kid** [comment]
```2.2.2022 20:11:39 [Error] [campaigncartographer] An exception was thrown when trying to start the mod:
System.NullReferenceException: Ссылка на объект не указывает на экземпляр объекта.
   в System.Object.GetType()
   в ApacheTech.Common.DependencyInjection.ServiceCollection.RegisterSingleton[TService](TService instance)
   в ApacheTech.VintageMods.Core.Hosting.DependencyInjection.Registration.ServerApiRegistrar.RegisterServerApiEndpoints(IServiceCollection container)
   в ApacheTech.Common.DependencyInjection.Extensions.HostExtensions.Configure(IServiceCollection services, Action`1 serviceRegistrationFactory)
   в ApacheTech.VintageMods.Core.Hosting.ModHost.BuildServerHost(ICoreServerAPI sapi)
   в ApacheTech.VintageMods.Core.Hosting.ModHost.StartPre(ICoreAPI api)
   ```
   List of files
   ```
02.02.2022  20:17    <DIR>          .
02.02.2022  20:17    <DIR>          ..
30.01.2022  19:57             1 929 2xArmorDurability_v1.0.0.zip
30.01.2022  18:54         1 522 558 alchemy-1.5.8.zip
30.01.2022  19:58           232 767 AlphaWeaponPack_v1.2.4.zip
30.01.2022  19:58           544 121 AnMiTech_v1.0.7.zip
30.01.2022  15:02           177 194 BackpackPackStandard_v1.0.3.zip
30.01.2022  18:49         1 943 455 BetterDrifters_v2.1.3.zip
30.01.2022  19:09         2 253 782 bricklayers-0.8.1.zip
30.01.2022  18:54            41 709 Bullseye 2.2.4.zip
30.01.2022  15:02            32 416 CarryCapacity-VS1.16.0-v0.6.4.zip
31.01.2022  19:40           610 598 cats_v1.3.0_petai_v1.3.x.zip
30.01.2022  18:36           417 788 CaveContent_v1.0.0.zip
30.01.2022  17:39            16 957 EasilyBalanceCrucibleCapacity-v1.0.0.zip
30.01.2022  16:24           210 918 ExpandedLiquidContainers-v1.1.0.zip
30.01.2022  18:38            14 372 ExtraOverlays_v1.16.0-pre.8_v1.1.0.zip
30.01.2022  17:39           456 584 FancyPlanters-v1.0.2.zip
30.01.2022  17:40             2 853 FixBadRecipes-v1.0.1.zip
30.01.2022  16:25            37 992 FixHandbookClutter-v1.0.1.zip
30.01.2022  19:01           129 405 FromGoldenCombs-1.16.0_v1.3.6-rc.3.zip
30.01.2022  18:37           153 198 hudclock-2.7.0.zip
30.01.2022  15:02         1 365 482 medievalexpansion-3.7.1.zip
31.01.2022  19:06           398 228 MoreClasses_v1.2.4.zip
30.01.2022  19:08             1 829 MoreDrifterLoot-1.5.1.zip
30.01.2022  16:22           155 501 MorePiles-v1.2.0.zip
30.01.2022  16:25           175 042 MorePilesQuasi-v1.0.1.zip
30.01.2022  15:02             1 357 MoreResin_1.0.zip
30.01.2022  18:48           312 525 necessaries_v1.16.0-1.1.1.zip
30.01.2022  17:39           113 875 PlaceEveryItem-v1.2.1.zip
30.01.2022  17:41            56 077 PlaceEveryItemFromMods-v1.0.1.zip
30.01.2022  15:03            25 758 PlayerCorpse_v1.16.0-pre.8_v1.3.0.zip
31.01.2022  19:35         6 501 193 qptech1.6.4.zip
30.01.2022  19:59           291 258 RangedWeapons-0.4.5.zip
30.01.2022  19:02            20 089 ResinRenaissance1.0.1-Fixed.zip
30.01.2022  17:42            21 734 RuinVariants-v1.0.0.zip
30.01.2022  19:03            12 983 starterpack-2.2.0.zip
30.01.2022  16:23            12 290 StoneBrickRebalance-v1.0.0.zip
30.01.2022  16:22            71 158 StonePathVariants-v1.0.2.zip
30.01.2022  18:50            96 617 StoneQuarry_v1.16.1_v1.6.0.zip
30.01.2022  18:47            13 289 survivalcats_v1.16.0-1.2.2-rc.2.zip
30.01.2022  18:41         3 828 868 TemporalTinkerer_v2.3.4.zip
30.01.2022  18:55            91 139 usefulstuff_v2.2.3.zip
30.01.2022  16:22           182 438 VanillaVariants-v1.1.2.zip
30.01.2022  19:19           115 200 VSCreativeMod.dll
30.01.2022  19:19           476 672 VSEssentials.dll
30.01.2022  19:19         1 809 920 VSSurvivalMod.dll
30.01.2022  18:55           269 925 wildcraft-v1.1.4.zip
30.01.2022  18:55            64 546 wildfarming_v1.1.4.zip
30.01.2022  18:02             4 720 wolffix_v1.0.0.zip
31.01.2022  19:40           509 716 wolftaming_v1.3.0_petai_v1.3.x.zip
30.01.2022  16:24           165 279 WoodBarrels-v1.0.1.zip
30.01.2022  17:40           631 875 WoodChests-v1.0.2.zip
30.01.2022  19:10           154 042 workbench-expansion-1.4.1.zip
30.01.2022  18:51            99 620 xlib_v0.6.8.zip
30.01.2022  18:51           217 369 xskills_v0.6.8.zip
```

**Comment by ApacheTech** [comment]
**Developer Notes:**

```csharp
        public static void RegisterServerApiEndpoints(IServiceCollection container)
        {
            container.RegisterSingleton(ApiEx.Server);
            container.RegisterSingleton(ApiEx.Server.World);
            container.RegisterSingleton(ApiEx.ServerMain);
            container.RegisterSingleton((ICoreAPICommon)ApiEx.Current);
            container.RegisterSingleton(ApiEx.Current);
        }
```

This seems, again, to be an edge-case issue, arising from the `ApiEx.Side` algorithm. This should (hopefully) be fixed within 2.1.0. By the time this method is called, the ApiEx end-points are all populated[1]; and Server.World will have been populated by the game... after all, that is the class that calls the mod loading procedure in the first place. The only place the method could fall over is when we reach `ApiEx.Current`, which uses `ApiEx.Side` to determine the app-side[2].

[1] See: https://github.com/ApacheTechSolutions/ApacheTech.VintageMods.Core/blob/master/ApacheTech.VintageMods.Core/Hosting/ModHost.cs#L72

[2] See: https://github.com/ApacheTechSolutions/ApacheTech.VintageMods.Core/blob/master/ApacheTech.VintageMods.Core/Common/StaticHelpers/ApiEx.cs#L115

**Comment by ApacheTech** [comment]
(Hopefully) Fixed in 2.1.0.
